platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbDataSource.java

```diff
@@ -1218,7 +1211,11 @@ private void writeLeavesToPathToKeyValue(
             // dirtyLeaves stream above
             if (isLongKeyMode) {
                 final long key = ((VirtualLongKey) leafRecord.getKey()).getKeyAsLong();
-                longKeyToPath.putIfEqual(key, path, INVALID_PATH);
+                if (isReconnect) {
+                    longKeyToPath.putIfEqual(key, path, INVALID_PATH);
+                } else {
+                    longKeyToPath.put(key, INVALID_PATH);
+                }
```

Comment:
Just curious; it looks like this is the essential fix.
What behavior did we uncover that reconnect should only replace a known/matching path with Invalid; while in other cases we replace or add a path as Invalid, regardless of previous value?

Reply:
As `isReconnect` flag suggests, there are two possible contexts:
- Regular flush. There we know for sure that if a tree node got into `deletedLeaves` set, it is deleted for sure and, therefore, the index should be updated unconditionally
- Flush on reconnect. In this case things may get tricky if a path of the key in question changed. In this case we delete conditionally (only if there were no changes), because if the path changed, it's a valid new path and should not be deleted.